packages/salesforce-adapter/src/filters/foreign_key_references.ts

```diff
@@ -0,0 +1,126 @@
+import { values, collections } from '@salto-io/lowerdash'
+import { FilterCreator } from '../filter'
+import { groupByAPIName, ApiNameMapping } from './instance_references'
+import { FOREIGN_KEY_DOMAIN } from '../constants'
+import { apiName } from '../transformers/transformer'
+
+const { makeArray } = collections.array
+
+/**
+ * Convert foreign-key-domain annotations to reference expressions using the known metadata types.
+ *
+ * @param elements      The fetched elements
+ * @param typeToElemID  Known element ids by metadata type
+ */
+const convertAnnotationsToReferences = (
+  elements: Element[],
+  typeToElemID: Record<string, ElemID>,
+): void => {
+  const resolveTypeReference = (ref: string | ReferenceExpression):
+    string | ReferenceExpression => {
+    if (_.isString(ref)) {
+      const referenceElemId = typeToElemID[ref]
+      if (referenceElemId !== undefined) {
+        return new ReferenceExpression(referenceElemId)
+      }
+    }
+    return ref
+  }
+
+  elements
+    .filter(isObjectType)
+    .flatMap((obj: ObjectType) => Object.values(obj.fields))
+    .filter((field: Field) => field.annotations[FOREIGN_KEY_DOMAIN] !== undefined)
+    .forEach((field: Field): void => {
+      field.annotations[FOREIGN_KEY_DOMAIN] = field.annotations[FOREIGN_KEY_DOMAIN]
+        .map(resolveTypeReference)
+    })
+}
+
+/**
+ * Resolve references using the mapping genrated from the Salesforce DescribeValueType API.
+ *
+ * @param instance                The current instance being modified
+ * @param apiNameToElemIDs        Known element ids, mapped by API name and metadata type
+ */
+const resolveReferences = (
+  instance: InstanceElement,
+  apiNameToElemIDs: ApiNameMapping,
+): void => {
+  const shouldReplace = (field: Field): boolean => (
+    makeArray(field.annotations[FOREIGN_KEY_DOMAIN]).some(isReferenceExpression)
+  )
+
+  const replacePrimitive = (val: Value, field: Field): Value => {
+    if (val === undefined || !_.isString(val)) {
+      return val
+    }
+
+    const refTarget = field.annotations[FOREIGN_KEY_DOMAIN]
+      .filter(isReferenceExpression)
+      .map((ref: ReferenceExpression) => apiNameToElemIDs[val]?.[
+        ref.elemId.typeName
+      ])
+      .find(values.isDefined)
+    return refTarget !== undefined ? new ReferenceExpression(refTarget) : val
+  }
+
+  const transformPrimitive: TransformFunc = ({ value, field }) => (
+    field !== undefined && shouldReplace(field) ? replacePrimitive(value, field) : value
+  )
+
+  // not using transformElement because we're editing the instance in-place
+  instance.value = transformValues({
+    values: instance.value,
+    type: instance.type,
+    transformFunc: transformPrimitive,
+    strict: false,
+  }) ?? instance.value
+}
+
+export const apiNameToElemID = (elements: Element[]): Record<string, ElemID> => (
+  _(elements)
+    .filter(isObjectType)
+    .map(e => [apiName(e), e.elemID])
+    .fromPairs()
+    .value()
```

Comment:
we generally prefer using built in language features when they are available outside lodash.
at a lower priority we are also trying to avoid using `_(...)` when it doesn't have a clear readability benefit for bundle size reasons (when using webpack it forces all of lodash to be included in the bundle whereas using `_.someFunc(...)` requires only the specific function to be in the bundle), this is a relatively new thing we are trying to do so there are still a lot of places in the code where you'll find `_(...)` but we do try to avoid introducing new ones so that when we eventually attempt to clean this up we'll have an easier time.

The exact same code can be written without lodash (and actually be shorter):
```suggestion
  Object.fromEntries(
    elements
      .filter(isObjectType)
      .map(e => [apiName(e), e.elemID])
  )
```
or, if we don't dive into `elemID` this can be the more common
```
  _.keyBy(elements.filter(isObjectType), e => apiName(e))
```